framework/src/Volo.Abp.AspNetCore.Authentication.OpenIdConnect/Microsoft/Extensions/DependencyInjection/AbpOpenIdConnectExtensions.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth.Claims;
+using Microsoft.AspNetCore.Authentication.OpenIdConnect;
+using Microsoft.Extensions.Options;
+using Volo.Abp.AspNetCore.MultiTenancy;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static class AbpOpenIdConnectExtensions
+    {
+        public static AuthenticationBuilder AddAbpOpenIdConnect(this AuthenticationBuilder builder)
+            => builder.AddAbpOpenIdConnect(OpenIdConnectDefaults.AuthenticationScheme, _ => { });
+
+        public static AuthenticationBuilder AddAbpOpenIdConnect(this AuthenticationBuilder builder, Action<OpenIdConnectOptions> configureOptions)
+            => builder.AddAbpOpenIdConnect(OpenIdConnectDefaults.AuthenticationScheme, configureOptions);
+
+        public static AuthenticationBuilder AddAbpOpenIdConnect(this AuthenticationBuilder builder, string authenticationScheme, Action<OpenIdConnectOptions> configureOptions)
+            => builder.AddAbpOpenIdConnect(authenticationScheme, OpenIdConnectDefaults.DisplayName, configureOptions);
+
+        public static AuthenticationBuilder AddAbpOpenIdConnect(this AuthenticationBuilder builder, string authenticationScheme, string displayName, Action<OpenIdConnectOptions> configureOptions)
+        {
+            return builder.AddOpenIdConnect(authenticationScheme, displayName, options =>
+            {
+                options.ClaimActions.MapAbpClaimTypes();
+
+                options.Events = new OpenIdConnectEvents
+                {
+                    OnAuthorizationCodeReceived = receivedContext =>
```

Comment:
I don't know that, but see it can be risky (because we are trusting a plain text value that can be changed by clients).

* When OnAuthorizationCodeReceived is called? In every request? While obtaining the first token by the client?
* What SetParameter does? Where this parameter goes and how it is used then?
* ...

Can you please explain this in details, I need to be sure that is a true approach.

Thanks.

Reply:
1. It will be called every time when the client receives the authorization code returned by the authorization server: https://github.com/dotnet/aspnetcore/blob/bbf7c8780c42e3d32aeec8018367037784eb9181/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L639
2. In code flow, the client will use the authorization code to send a request to get an access token. https://github.com/dotnet/aspnetcore/blob/bbf7c8780c42e3d32aeec8018367037784eb9181/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L807v
   This parameter will be serialized as `FormUrlEncodedContent` as a request parameter, but it missing `TenantKey`, so I use the `SetParameter` method to add the `TenantKey` parameter.  (actually, `Tenantkey` is in the cookie, but due to the same-origin policy, the browser will not send the cookie)